Fixes part of #2322: Changes in Previous Responses Header Item [A11y]

Mehul-Bisht opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
  
Fixes part of #2322: Changes in Previous Responses Header Item [A11y]  
This PR fixes the accessibility issue for the height of Previous Responses Header in mobile layouts only, since I was not able to set up the accessibility scanner on my tablet emulator correctly.

## Screenshots of UI
| Before | After |
:-------------------------:|:-------------------------:
![header_before](https://user-images.githubusercontent.com/65572088/107808191-314fdb00-6d8f-11eb-94fe-d12b771ab4e1.png) | ![header_after](https://user-images.githubusercontent.com/65572088/107808229-42005100-6d8f-11eb-8ba2-f6046b6f2ce8.png)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Mehul-Bisht committed:
fixed issue for mobile layouts

Mehul-Bisht committed:
reverted changes for .idea/misc.xml file

Mehul-Bisht commented:
@ARJUPTA @rt4914 kindly review the changes

Arjupta reviewed (changes requested):
@Mehul-Bisht -
- Refer the issue number in the description
- Screenshot of UI before and after the changes . Also for the tests that are passing

Can you explain the which part of the issue did you tried to solve. because there is only one problem mentioned in the #2322 ie of small clickable region

Mehul-Bisht commented:
> Can you explain the which part of the issue did you tried to solve. because there is only one problem mentioned in the #2322 ie of small clickable region

@ARJUPTA the previous responses header item had an accessibility issue with its touch target height that I fixed in this PR for mobile layouts, but not for the tablet layout. I have also updated the PR description.

Arjupta commented:
> > Can you explain the which part of the issue did you tried to solve. because there is only one problem mentioned in the #2322 ie of small clickable region
> 
> @ARJUPTA the previous responses header item had an accessibility issue with its touch target height that I fixed in this PR for mobile layouts, but not for the tablet layout. I have also updated the PR description.

Okay taking a look. Meanwhile can you resolve the conflict

Mehul-Bisht commented:
> Okay taking a look. Meanwhile can you resolve the conflict

The conflict seems not to be a result of changes made in this PR. Is there anything you can suggest that I may try?

Arjupta reviewed (commented):
@rt4914 The results are like this for the mobile layout
|Before|After|
|:-:|:-:|
|![Screenshot_1613240390](https://user-images.githubusercontent.com/64064110/107868328-9c8dd000-6ea9-11eb-8921-b81705def224.png)|![Screenshot_1613240871](https://user-images.githubusercontent.com/64064110/107868333-9f88c080-6ea9-11eb-8e9e-cd5a99526fcd.png)|
|![Screenshot_1613240453](https://user-images.githubusercontent.com/64064110/107868330-9e579380-6ea9-11eb-819e-c404ed6718f4.png)|![Screenshot_1613240845](https://user-images.githubusercontent.com/64064110/107868332-9ef02a00-6ea9-11eb-9854-8a16ceab57ab.png)|

but do we need to change the padding dimension to what @Mehul-Bisht suggested or we should keep earlier ones as they were more descriptive

Arjupta commented:
> > Okay taking a look. Meanwhile can you resolve the conflict
> 
> The conflict seems not to be a result of changes made in this PR. Is there anything you can suggest that I may try?

The conflict is about having duplicate dimensions, try to merge your code locally to develop and then see the files which show conflict. You can learn about resolving conflicts from any video.

Mehul-Bisht committed:
resolved conflicts

Mehul-Bisht commented:
@ARJUPTA I've resolved the conflicts, kindly discuss with Rajat and let me know about the final decision

rt4914 commented on app/src/main/res/values/dimens.xml:
```diff
@@ -176,9 +176,9 @@
+  <dimen name="previous_responses_item_exploration_view_margin_top">0dp</dimen>
   <dimen name="previous_responses_item_question_split_view_margin_top">24dp</dimen>
-  <dimen name="previous_responses_item_question_view_margin_top">24dp</dimen>
+  <dimen name="previous_responses_item_question_view_margin_top">0dp</dimen>
```
If there is a padding / margin of `0dp` remove that and use `space_0dp` where it was used in code.

rt4914 replied:
Also merge with latest develop

Mehul-Bisht replied:
@rt4914 Ok, and should I let this dimension be 0dp now or revert it back to 24dp?

Arjupta replied:
@rt4914 so are we following that `space_0dp` suggestion which we concluded while renaming all the dimensions during SLOP

Arjupta replied:
> @rt4914 Ok, and should I let this dimension be 0dp now or revert it back to 24dp?

You just have to use space_0dp in place of a named dimension with a value of 0dp. So don't revert back.

rt4914 replied:
Basically all dimens values should be of some value other than 0dp and whereever you want to use `0dp` use `space_0dp`.

rt4914 replied:
This does not seem to be solved yet.

rt4914 reviewed (changes requested):
@Mehul-Bisht PTAL.

Mehul-Bisht committed:
Merge remote-tracking branch 'upstream/develop' into changes-in-previous-responses-header-items

Mehul-Bisht committed:
updated margins to space_0dp

Mehul-Bisht committed:
reverted changes for .idea/misc file

Arjupta reviewed (changes requested):
@Mehul-Bisht as `previous_responses_item_exploration_view_margin_top` and `previous_responses_item_question_view_margin_top` are replaced with space_0dp, hence they can be removed from dimens.xml

Mehul-Bisht commented:
>@Mehul-Bisht as previous_responses_item_exploration_view_margin_top and previous_responses_item_question_view_margin_top are replaced with space_0dp, hence they can be removed from dimens.xml

@ARJUPTA though i have not made changes in this PR for tablet layouts, should I still consider removing them from the mobile layouts?

Arjupta commented:
> > @Mehul-Bisht as previous_responses_item_exploration_view_margin_top and previous_responses_item_question_view_margin_top are replaced with space_0dp, hence they can be removed from dimens.xml
> 
> @ARJUPTA though i have not made changes in this PR for tablet layouts, should I still consider removing them from the mobile layouts?

Then it will better if you include the similar changes for a Tablet Layout too and ask us for a review if you are not able to test it

BenHenning committed:
Merge branch 'develop' into changes-in-previous-responses-header-items

Mehul-Bisht committed:
fixed issue with tablet layout in portrait mode

Mehul-Bisht committed:
Merge remote-tracking branch 'upstream/develop' into changes-in-previous-responses-header-items

Mehul-Bisht committed:
Merge branch 'changes-in-previous-responses-header-items' of https://github.com/Mehul-Bisht/oppia-android into changes-in-previous-responses-header-items

Mehul-Bisht committed:
reverted changes for idea/misc

Mehul-Bisht commented:
## Update:

## Portrait Mode for Tablet:
| Before |
:-------------------------:
![screenshot-2021-03-01_03 15 51 599](https://user-images.githubusercontent.com/65572088/109435345-16908e00-7a40-11eb-9da3-84dbca702a6a.png)
| After |
![screenshot-2021-03-01_03 29 01 176](https://user-images.githubusercontent.com/65572088/109435361-2ad48b00-7a40-11eb-9245-adeea5c8f18b.png)

## Landscape Mode for Tablet:

There is no issue with landscape mode for tablet in the emulator that I used, here is the result for me
![screenshot-2021-03-01_03 16 29 19](https://user-images.githubusercontent.com/65572088/109435402-548db200-7a40-11eb-9b8e-ae74bfc7944b.png)

Mehul-Bisht commented:
@ARJUPTA @rt4914 Please check the changes for tablet layouts, I was able to setup the emulator for tablet.

Arjupta commented on app/src/main/res/values-sw600dp-port/dimens.xml:
```diff
@@ -54,6 +54,7 @@
   <dimen name="previous_responses_item_exploration_view_margin_top">24dp</dimen>
   <dimen name="previous_responses_item_question_split_view_margin_top">24dp</dimen>
   <dimen name="previous_responses_item_question_view_margin_top">24dp</dimen>
+  <dimen name="previous_response_header_padding_top">24dp</dimen>
```
As these values are not different from the ones in `values/dimens.xml` so they are not needed to be repeated here. Also we should remove the top_margins as they are already set to be space_0dp

Arjupta reviewed (changes requested):
@Mehul-Bisht

Mehul-Bisht committed:
removed unused top margins

Mehul-Bisht commented:
@ARJUPTA PTAL now

rt4914 reviewed (changes requested):
@Mehul-Bisht Can you please add before and after screenshot with exact combination as a proof that this PR does not change any margins in overall. Mobile portrait, landscape tablet portrait and landscape.

Mehul-Bisht committed:
resolved conflicts

Mehul-Bisht commented:
## Mobile mode

##  Portrait mode
| Before | After |
:-------------------------:|:-------------------------:
![mobile_port_before](https://user-images.githubusercontent.com/65572088/110559919-15134400-816b-11eb-9dca-c4ab22ca6136.png) | ![mobile_port_after](https://user-images.githubusercontent.com/65572088/110559953-25c3ba00-816b-11eb-9d77-7a6010ae8843.png)

## Landscape mode
| Status |
:-------------------------:
## Before 
![mobile_landscape_before](https://user-images.githubusercontent.com/65572088/110560031-48ee6980-816b-11eb-91d3-7638f4629410.png)
## After 
![mobile_landscape_after](https://user-images.githubusercontent.com/65572088/110560050-5146a480-816b-11eb-9efc-c318fd27bd05.png)

## Tablet mode

##  Portrait mode
| Status | 
:-------------------------:
## Before
![tablet_port_before](https://user-images.githubusercontent.com/65572088/110560089-61f71a80-816b-11eb-83fe-efae52b5575a.png) 

## After
![tablet_port_after](https://user-images.githubusercontent.com/65572088/110560096-66bbce80-816b-11eb-9772-d072a257c9eb.png)

## Landscape mode
| Status |
:-------------------------:
## Before 
![tablet_land_before](https://user-images.githubusercontent.com/65572088/110560130-71766380-816b-11eb-8504-46575d5192fc.png)

## After
![tablet_land_after](https://user-images.githubusercontent.com/65572088/110560148-7d622580-816b-11eb-9560-ee714d64d968.png)

Mehul-Bisht commented:
> @Mehul-Bisht Can you please add before and after screenshot with exact combination as a proof that this PR does not change any margins in overall. Mobile portrait, landscape tablet portrait and landscape.

@rt4914 PTAL

rt4914 commented:
@Mehul-Bisht There is some issue with this PR because of which it is showing 100+ files changed and also checks are not running. Can you try merging with develop again?

Mehul-Bisht committed:
Merge remote-tracking branch 'upstream/develop' into changes-in-previous-responses-header-items

Mehul-Bisht commented:
@rt4914 I had merge conflicts but I was unable to view them in my local branch, so I pulled changes into develop and got changes from latest develop. Kindly PTAL, the checks have started running now.

rt4914 commented:
@Mehul-Bisht The PR looks good but you have missed one change. Please take a look at this: https://github.com/oppia/oppia-android/pull/2703/files#r576944928

Mehul-Bisht commented:
@rt4914 I had a doubt, if there is a dimension that is being used by one of the tablet orientations only and not by the other 3 (mobile landscape and portrait and the other tablet orientation) then how should I be declaring it for that single orientation? Since the IDE warns me with `dimen has no declaration in the base values folder; this can lead to crashes when the resource is queried in a configuration that does not match this qualifier`. For other orientations it will be using `space_0dp` and hence does not require to be declared in the base values folder.

rt4914 commented:
> @rt4914 I had a doubt, if there is a dimension that is being used by one of the tablet orientations only and not by the other 3 (mobile landscape and portrait and the other tablet orientation) then how should I be declaring it for that single orientation? Since the IDE warns me with `dimen has no declaration in the base values folder; this can lead to crashes when the resource is queried in a configuration that does not match this qualifier`. For other orientations it will be using `space_0dp` and hence does not require to be declared in the base values folder.

@Mehul-Bisht I recently encountered this issue and therefore we will need to create the same name in `values/dimens.xml` file too with `0dp` value.

anandwana001 commented:
@Mehul-Bisht any update on this?

Mehul-Bisht commented:
> @Mehul-Bisht any update on this?

Yes, I will be making the suggested changes as soon as possible

Mehul-Bisht commented:
@rt4914 I seem to have lost track of what that issue here was, regarding naming dimens in all files to `0dp` which have a `0dp` value instead of some value. I think the portrait mode for tablet doesn't have `previous_responses_item_exploration_view_margin_top` set to `0dp` yet. Was that the issue? 

Also, since I'm building from VirtualBox Ubuntu I won't be able to test for tablets, only on my real device (mobile). Kindly someone test changes for tablet layouts.

rt4914 commented:
@Mehul-Bisht You can leave this issue. As I was working on this screen it became evident that UI depends upon other items too so its not that straight forward. Closing this now.
